### PR TITLE
Spec ecosystem incentive

### PR DESCRIPTION
**Pull request by taiki1frsy**
This PR is for the revision of the previous frontend-incentive module specification after the research about the things on check_list in there.
This PR is merged into the PR #216 . After merged, that PR also is merged into `design/spec` branch manually.

The changed parts from the previous frontend-incentive specification are not a few actually.

About module name
I changed module name from frontend-incentive to ecosystem-incentive because the current style can be applied for many different use-cases by using the same logic.

About logic
The biggest modification is definitely hooks part.
As you know, in the previous design, we planed to use event data for catching the traded fee information to accumulate rewards for the subjects.
But, after the research about it, it turned out to be impossible to do so.
Instead, there're hooks mechanism for cosmos sdk for literally hooking to some target actions.
(research log: https://www.notion.so/cauchye/Frontend-incentive-Tech-research-9b0965c91cd542ecbb6ea123d557add3)
By using this mechanism and the way to get memo data in a specific message process in the how_get_memo branch, we can implement incentive features without generating two-way dependencies between x/nftmarket and x/ecosystem-incentive modules.

And, following these changes, I decided to accept register style to become subject.

refs:
https://github.com/UnUniFi/chain/issues/218
https://www.notion.so/cauchye/Frontend-incentive-Tech-research-9b0965c91cd542ecbb6ea123d557add3

Please review and confirm, @mkXultra .

### CODE REVIEW

**Comment by taiki1frsy** [comment]
SetHooks usage in app.go: https://github.com/UnUniFi/chain/blob/a9f86eb184d518bd5122a5ad5a98d109f791ac6d/app/app.go#L461
https://github.com/UnUniFi/chain/blob/a9f86eb184d518bd5122a5ad5a98d109f791ac6d/app/app.go#L555

The location to define `SetHooks` in x/cdp/keeper/keeper.go: https://github.com/UnUniFi/chain/blob/a9f86eb184d518bd5122a5ad5a98d109f791ac6d/x/cdp/keeper/keeper.go#L57

File: x/ecosystem-incentive/spec/02_state.md [hunk 1]
```diff
@@ -2,24 +2,24 @@
...
+- Incentive: `"incentive_id" -> format(IncentiveStore)`
 
 ## RewardTable
 
-RewardTable is the record of the rewards for the subject of the `frontend-incentive`.
+RewardTable is the record of the rewards for the subject of the `ecosystem-incentive`.
 
 ```protobuf
-message FrontendReward {
+message Reward {
```
Comment (mkXultra) [inline x/ecosystem-incentive/spec/02_state.md:22]: @taiki1frsy 
I think the data structure is ambiguous.
Can I see a sample data structure?

Comment (taiki1frsy) [reply x/ecosystem-incentive/spec/02_state.md:38]: @mkXultra 
I added description in 02_state.md for more abundant document of the data structures in this module.
Plz refer.

File: x/ecosystem-incentive/spec/02_state.md [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+  repeated undetermined weights = 3;
+  RewardType reward_type = 4;
+}
+```
+
+- Incentive: `"incentive_id" -> format(IncentiveStore)`
+
+## IncentiveIdTable
+
+- incentive_id_table: `format(nft_id) -> format(incentive_id)`
```
Comment (mkXultra) [inline x/ecosystem-incentive/spec/02_state.md:29]: what is this?
why we use nft if

**Comment by mkXultra** [comment]
It might be better to have a Keeper Method that punishes it if there is a malicious front end.
It can be made to run with ProposalOnly

**Comment by taiki1frsy** [comment]
TODO:
- Add hook method to trigger the delete data about nft_id - incentive_id
- Define the locations to call above method

**Comment by taiki1frsy** [comment]
Memo:
Currently we don't have liquidation, auction, etc logic in the code.
So when we implement those stuff, the hook insertion must be included in there.

File: x/ecosystem-incentive/spec/02_state.md [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+
+The v1 archtecture is:
+
+```json
+{
+  "version": "v1",
+  "incentive-type": 0,
+  "incentive-id": "incentive_id"
+}
+```
```
Comment (mkXultra) [inline x/ecosystem-incentive/spec/02_state.md:100]: Please create a new document for this part.
There is a problem with stating in the state.md

File: x/ecosystem-incentive/spec/04_hooks.md [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+
+**NOTE: This is early draft.**
+
+All rewards accumulation are executed when the according hooks function is called.   
+
+The example hooks functions interfaces in x/nftmarket module:
+
+```go
+type NftmarketHooks interface {
+   AfterNftListed(ctx sdk.Context, nft_id types.NftIdentifier, incentive_id string)
```
Comment (mkXultra) [inline x/ecosystem-incentive/spec/04_hooks.md:11]: Put tx.memo in the argument.
incentive_id is not required
```
before
AfterNftListed(ctx sdk.Context, nft_id types.NftIdentifier, incentive_id string)
after
AfterNftListed(ctx sdk.Context, nft_id types.NftIdentifier, tx_memo string)
```

**Review by mkXultra — CHANGES_REQUESTED** [review]

File: x/ecosystem-incentive/spec/02_state.md (hunk 3 above)
Comment (taiki1frsy) [reply x/ecosystem-incentive/spec/02_state.md]: Addressed at https://github.com/UnUniFi/chain/pull/225/commits/506cc07198e8f79dac3b27f4c6800a3fd34c6500.

File: x/ecosystem-incentive/spec/04_hooks.md (hunk 4 above)
Comment (taiki1frsy) [reply x/ecosystem-incentive/spec/04_hooks.md]: Modified as it shows.
https://github.com/UnUniFi/chain/blob/23b8c63c40c112e6adceb0c40ae492c5ba8067b5/x/ecosystem-incentive/spec/04_hooks.md?plain=1#L11

**Review by mkXultra — APPROVED** [review]

**Pull request merged**